c10/core/impl/SizesAndStrides.h

```diff
@@ -8,6 +10,17 @@ namespace c10 {
 namespace impl {
 
 // Packed container for TensorImpl sizes and strides.
+// This design improves on the previous approach of using a pair of
+// c10::SmallVector<int64_t, 5> by specializing for the operations we
+// actually use and enforcing that the number of sizes is the same as
+// the number of strides. The memory layout is as follows:
+//
+// 1 tagged intptr_t; if the tag bit is 1, the data is stored inline and
+// the rest of the word indicates the size of the data. If the tag bit
+// is 0, the tag word is a pointer to an array of int64_t holding the
+// data; the first int64_t indicates the size of the rest.
+// 5 int64_t reserved for inline sizes
+// 5 int64_t reserved for inline strides
```

Comment:
I liked the tagged bit proposal when you first proposed it, but after taking a closer look at this PR, I was thinking: why do we need a tagged pointer at all? Make the first word the size unconditionally; when size is greater than 5, stash the pointer in the inline fields. When we store out of line the inline space is totally wasted, so might as well just use it to store the pointer. (You could even inplace construct the vector here, so you don't have to reimplement an out-of-line constructor).

The benefit of the alternate encoding is that you can now retrieve the dim() of a tensor without any branches. The net number of branches stays the same. I guess one downside is some of the bookkeeping is more complicated since you have to avoid clobbering data you might need for later in the inline fields.

Reply:
> Make the first word the size unconditionally; when size is greater than 5, stash the pointer in the inline fields. 

This gives us two places that store the size for the out-of-line case, adding implementation complexity. However, see the second benefit below!

> The benefit of the alternate encoding is that you can now retrieve the dim() of a tensor without any branches. 

I don't think that this benefit is worthwhile. A correctly predicted branch takes 1 cycle, and we expect much more than 99% of tensors to have rank 5 or below. We should expect a correct prediction on this the vast majority of the time.

That said, I can think of some other benefits you don't mention:
- We get to drop the bit shift to retrieve the size, which is nice to have and simplifies other code paths.
- Optimizing the out-of-line representation is a lot easier -- instead of a pointer to a vector, we can just use a pointer to an array of the right size directly, without worrying about where to put the size.

Overall, I think this is a good idea. After I'm sure that test coverage for the current version is good, I'll switch over to this suggestion and get rid of the `vector` as well. Thanks!